File: src/main/java/raccoonlang/parser/ast/ModifiersAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.Token;

public class ModifiersAstNode {
    public boolean isPublic;
    public boolean isPrivate;
    public boolean isProtected;
    public boolean isInternal;

    public static ModifiersAstNode parse(Parser parser) {
        ModifiersAstNode modifiersAstNode = new ModifiersAstNode();

        boolean wasModifier = true;
        while (wasModifier){
            Token token = parser.peek();
            switch (token.getTokenType()){
                case PUBLIC:
                    if(modifiersAstNode.isPublic) throw new RuntimeException("Duplicate modifier at: " + token);// TODO: custom exception
                    modifiersAstNode.isPublic = true;
                    break;
                case PRIVATE:
                    if(modifiersAstNode.isPrivate) throw new RuntimeException("Duplicate modifier at: " + token);// TODO: custom exception
                    modifiersAstNode.isPrivate = true;
                    break;
                case PROTECTED:
                    if(modifiersAstNode.isProtected) throw new RuntimeException("Duplicate modifier at: " + token);// TODO: custom exception
                    modifiersAstNode.isProtected = true;
                    break;
                case INTERNAL:
                    if(modifiersAstNode.isInternal) throw new RuntimeException("Duplicate modifier at: " + token);// TODO: custom exception
                    modifiersAstNode.isInternal = true;
                    break;
                default:
                    wasModifier = false;
            }
        }

        return modifiersAstNode;
    }

    @Override
    public String toString() {
        return "ModifiersAstNode{" +
                "isPublic=" + isPublic +
                ", isPrivate=" + isPrivate +
                ", isProtected=" + isProtected +
                ", isInternal=" + isInternal +
                '}';
    }
}

File: Parser/AST/ModifiersAstNode.cs
namespace Raccoonlang.Parser.AST;

using Tokenizer;
using Exception;

public class ModifiersAstNode {
    bool isPublic;
    bool isPrivate;
    bool isProtected;
    bool isInternal;

    public static ModifiersAstNode Parse(Parser parser)
    {
        ModifiersAstNode node = new ModifiersAstNode();

        bool isMod = true;

        while (isMod) {
            Token t = parser.Peek();

            switch(t.Type) {
                case TokenType.PUBLIC:
                    if(node.isPublic) throw new DuplicateModifierException(t);
                    node.isPublic = true;
                    break;
                case TokenType.PRIVATE:
                    if(node.isPrivate) throw new DuplicateModifierException(t);
                    node.isPrivate = true;
                    break;
                case TokenType.PROTECTED:
                    if(node.isProtected) throw new DuplicateModifierException(t);
                    node.isProtected = true;
                    break;
                case TokenType.INTERNAL:
                    if(node.isInternal) throw new DuplicateModifierException(t);
                    node.isInternal = true;
                    break;
                default:
                    isMod = false;
                    break;
            }

            if(isMod) {
                parser.Take();
            }
        }

        return node;

    }

    public override string ToString()
    {
        return "ModifiersAstNode{" +
                "isPublic=" + isPublic +
                ", isPrivate=" + isPrivate +
                ", isProtected=" + isProtected +
                ", isInternal=" + isInternal +
                "}";
    }

}